Add spec and parser for luksmeta command

danzatt opened the pull request:
Signed-off-by: daniel.zatovic <daniel.zatovic@gmail.com>

### All Pull Requests:

Check all that apply:

* [x] Have you followed the guidelines in our Contributing document, including the instructions about commit messages?
* [ ] Is this PR to correct an issue?
* [x] Is this PR an enhancement?

### Complete Description of Additions/Changes:

- added `luks1_block_devices` which depends on `LuksDump` parser and filters just LUKS1 devices and returns their paths (based on the UUID)
- added `luksmeta` spec which runs `luksmeta show` command on every LUKS1 device to show information about custom metadata embedded in the keyslots
- added `LuksMeta` parser to parse the `luksmeta show` command

xiangce commented on insights/parsers/luksmeta.py:
```diff
@@ -0,0 +1,79 @@
+
+
+    Attributes:
+        keyslots(dict): A list of 8 KeyslotSpecification objects, describing
```
I think this `keyslots` should be a list.

xiangce replied:
BTW, if it works for you, it's recommended to mix the parser with the built-in `list` and remove this single attribute, when it's a simple list.


```
class LuksMeta(Parser, list):

    self[index] = ...

```

xiangce commented on insights/parsers/luksmeta.py:
```diff
@@ -0,0 +1,79 @@
+
+    Attributes:
+        keyslots(dict): A list of 8 KeyslotSpecification objects, describing
+        every LUKS keyslot. The KeyslotSpecification contains the index, state
```
To make sure the docstring can be shown correctly after being compiled, please add one more intention from the 2nd line, e.g.
```
keyslots(list): A list of 8 KeyslotSpecification objects, describing
      and metadata fileds. Metadata field stores the UUID of the application
      that has stored metadata in the keyslot.
```

xiangce commented on insights/parsers/luksmeta.py:
```diff
@@ -0,0 +1,79 @@
+    """  # noqa
+
+    def __init__(self, context):
+        self.keyslots = [None] * 8
```
Does it make sense to "skip" this parser by raising a `SkipException` when there is nothing needed to parse?

xiangce commented on insights/specs/datasources/luks_devices.py:
```diff
@@ -123,3 +124,28 @@ def luks_data_sources(broker):
     raise SkipComponent
+
+
+@datasource(LuksDump, HostContext)
```
Since the `LuksDump` is required for this datasource, please add it to the `insights/collect.py` too.

bfahr replied:
@xiangce @danzatt why not just use parsers and a collector instead of adding a datasource that has to run on a customer systems?

xiangce commented on insights/parsers/luksmeta.py:
```diff
@@ -0,0 +1,79 @@
+"""
+luksmeta - command ``luksmeta show``
+====================================
+This class provides parsing for the output of luksmeta <device_name>.
```
it should be `luksmeta -d <device_name>` here in the doc?

xiangce replied:
You may miss this one.

danzatt replied:
Fixed now

xiangce reviewed (changes requested):
I also find this [KCS-3617741](https://access.redhat.com/solutions/3617741), please check if this error `"Device is not initialized (/dev/mapper/vg_luks-lv_luks)"` should be handled or not in this parser. Thanks.

danzatt commented:
I have gone through the remarks and fixed them now.

xiangce reviewed (approved):
Thanks for the update, it looks good to me now.

bfahr commented on insights/parsers/luksmeta.py:
```diff
@@ -0,0 +1,98 @@
+
+        # LUKS1 contains exactly 8 keyslots
+        if len(content) != 8:
+            raise SkipComponent
```
@danzatt your comment implies that there must be 8 keyslots.  Is it an error if there are not exactly 8?

bfahr commented on insights/specs/datasources/luks_devices.py:
```diff
@@ -123,3 +124,28 @@ def luks_data_sources(broker):
     raise SkipComponent
+
+
+@datasource(LuksDump, HostContext)
```
@xiangce @danzatt why not just use parsers and a collector instead of adding a datasource that has to run on a customer systems?

xiangce replied:
@danzatt, yeap, what @bfahr suggested makes sense to me. It seems possible to run the `luksmeta -d` command against all the `/dev/disk/by-uuid/*` devices. And then get what you wanted (uuids where ["header"]["Version"] == "1" in LuksDump) via a combiner.

xiangce commented on insights/specs/default.py:
```diff
@@ -369,6 +370,7 @@ class DefaultSpecs(Specs):
     lspci_vmmkn = simple_command("/sbin/lspci -vmmkn")
     lsscsi = simple_command("/usr/bin/lsscsi")
     lsvmbus = simple_command("/usr/sbin/lsvmbus -vv")
+    luksmeta = foreach_execute(cryptsetup.luks1_block_devices, "luksmeta show -d %s")
```
BTW, please use the full command path for the `luksmeta` command (/usr/bin/luksmeta) in the spec, since the insights-client may not get the correct PATH environment.


And, Would you please modify the following two specs to use the full path (/usr/sbin/cryptsetup)? https://github.com/RedHatInsights/insights-core/blob/534888a20d32c256dd1443fe8b281c9927e4054d/insights/specs/datasources/luks_devices.py#L43-L44

And sorry for missing this in the review of that PR.

xiangce commented on insights/collect.py:
```diff
@@ -251,6 +251,9 @@
         - name: insights.components.cryptsetup
           enabled: true
+
+        - name: insights.combiners.cryptsetup.luks1_block_devices
```
If It's okay to run the `luksmeta` against all the `by-uuid` devices, this component is not required and is not needed to be `pre-loaded`.

danzatt commented:
Yes, indeed it should be simpler to run the `luksmeta` command against all devices. I have changed the PR now and removed the `luks1_block_devices` dependency. (I fixed the `cryptsetup`/`luksmeta` paths too.)

xiangce commented on insights/combiners/cryptsetup.py:
```diff
@@ -0,0 +1,80 @@
+from insights.parsers.luksmeta import LuksMeta
+
+
+@combiner(LuksDump)
```
@danzatt - from my perspective, It seems this combiner function is not so required. Firstly, it's based on the single parser, if really want such an interface to return the "luks1 block devices", you can totally implement it in the `LuksDump` parser itself. Or, you can move this check into the `LuksDevices` combiner as a double safety-check here L77: `if luks_metas and uuid in luksmeta_by_uuid:`

xiangce commented on insights/combiners/cryptsetup.py:
```diff
@@ -0,0 +1,80 @@
+            if luks_metas and uuid in luksmeta_by_uuid:
+                luks_dump_copy["luksmeta"] = luksmeta_by_uuid[uuid]
+
+            self.append(luks_dump_copy)
```
Better add a check here: when got nothing from the combination (e.g. no luks devices and the list is empty), better raise a SkipComponent to skip the combiner.

danzatt commented:
Is there anything I can do to get this merged?

xiangce commented on insights/combiners/cryptsetup.py:
```diff
@@ -0,0 +1,80 @@
+    return luks1_devices
+
+
+@combiner(LuksDump, optional=[LuksMeta])
```
I just checked your latest update to the test of this combiner, it now looks good to me. Sorry for the late reply, I was just back from a long leave.

danzatt replied:
Great, thank you for the review.
